Ensure initializedContextIds_ map is cleaned up appropriately in DistEngine.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29787 Ensure initializedContextIds_ map is cleaned up appropriately in DistEngine.**

The initializedContextIds_ map was never cleaned up in DistEngine and
kept on growing as we continue to run backward passes. To fix this, in this PR
we ensure that the context id is cleaned up from this map once we are done with
the backward pass.

Closes #29083

Differential Revision: [D18498937](https://our.internmc.facebook.com/intern/diff/D18498937/)

Commit:
Ensure initializedContextIds_ map is cleaned up appropriately in DistEngine.

The initializedContextIds_ map was never cleaned up in DistEngine and
kept on growing as we continue to run backward passes. To fix this, in this PR
we ensure that the context id is cleaned up from this map once we are done with
the backward pass.

Closes #29083

Differential Revision: [D18498937](https://our.internmc.facebook.com/intern/diff/D18498937/)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/autograd/engine/dist_engine.h:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <gtest/gtest_prod.h>
```
I don't like this... I know it's for making protected functions testable, but isn't there another way to achieve the same? For example, do the functions you're calling into need to be protected?

pietern reviewed (commented):
Not a fan of including the gtest header... 

Otherwise the change LGTM.

Commit:
Update on "Ensure initializedContextIds_ map is cleaned up appropriately in DistEngine."


The initializedContextIds_ map was never cleaned up in DistEngine and
kept on growing as we continue to run backward passes. To fix this, in this PR
we ensure that the context id is cleaned up from this map once we are done with
the backward pass.

Closes #29083

Differential Revision: [D18498937](https://our.internmc.facebook.com/intern/diff/D18498937/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -212,6 +212,7 @@ void DistEngine::executeSendFunction(
 
     // Mark the autograd context id as initialized and unlock.
     initializedContextIds_.insert(autogradContext.contextId());
+    ClearContextIdGuard guard(autogradContext.contextId());
```
why this needs to be before the unlock?

mrshenli reviewed (approved):
This LGTM, just a minor comment on where to create the guard.

Commit:
Update on "Ensure initializedContextIds_ map is cleaned up appropriately in DistEngine."


The initializedContextIds_ map was never cleaned up in DistEngine and
kept on growing as we continue to run backward passes. To fix this, in this PR
we ensure that the context id is cleaned up from this map once we are done with
the backward pass.

Closes #29083

Differential Revision: [D18498937](https://our.internmc.facebook.com/intern/diff/D18498937/)

[ghstack-poisoned]
